lib/remote_node.go

```diff
@@ -0,0 +1,648 @@
+}
+
+// setStatusTerminated sets the connection status of the remote node to terminated.
+func (rn *RemoteNode) setStatusTerminated() {
+	rn.connectionStatus = RemoteNodeStatus_Terminated
+}
+
+// setStatusAttempted sets the connection status of the remote node to attempted.
+func (rn *RemoteNode) setStatusAttempted() {
+	rn.connectionStatus = RemoteNodeStatus_Attempted
+}
+
+func (rn *RemoteNode) GetId() RemoteNodeId {
+	return rn.id
+}
+
+func (rn *RemoteNode) GetPeer() *Peer {
+	return rn.peer
+}
+
+func (rn *RemoteNode) GetNegotiatedProtocolVersion() ProtocolVersionType {
+	return rn.handshakeMetadata.negotiatedProtocolVersion
+}
+
+func (rn *RemoteNode) GetValidatorPublicKey() *bls.PublicKey {
+	return rn.handshakeMetadata.validatorPublicKey
+}
+
+func (rn *RemoteNode) GetUserAgent() string {
+	return rn.handshakeMetadata.userAgent
+}
+
+func (rn *RemoteNode) getHandshakeStage() HandshakeStage {
+	return rn.handshakeMetadata.handshakeStage
+}
+
+func (rn *RemoteNode) setHandshakeStage(stage HandshakeStage) {
+	rn.handshakeMetadata.handshakeStage = stage
+}
+
+func (rn *RemoteNode) IsInbound() bool {
+	return rn.peer != nil && !rn.peer.IsOutbound()
+}
+
+func (rn *RemoteNode) IsOutbound() bool {
+	return rn.peer != nil && rn.peer.IsOutbound()
+}
+
+func (rn *RemoteNode) IsPersistent() bool {
+	return rn.peer != nil && rn.peer.IsPersistent()
+}
+
+func (rn *RemoteNode) IsNotConnected() bool {
+	return rn.connectionStatus == RemoteNodeStatus_NotConnected
+}
+
+func (rn *RemoteNode) IsConnected() bool {
+	return rn.connectionStatus == RemoteNodeStatus_Connected
+}
+
+func (rn *RemoteNode) IsValidated() bool {
+	return rn.connectionStatus == RemoteNodeStatus_HandshakeCompleted
+}
+
+func (rn *RemoteNode) IsValidator() bool {
+	if !rn.IsValidated() {
+		return false
+	}
+	return rn.GetValidatorPublicKey() != nil
+}
+
+// DialOutboundConnection dials an outbound connection to the provided netAddr.
+func (rn *RemoteNode) DialOutboundConnection(netAddr *wire.NetAddress) error {
+	if !rn.IsNotConnected() {
+		return fmt.Errorf("RemoteNode.DialOutboundConnection: RemoteNode is not in the NotConnected state")
+	}
+
+	rn.mtx.Lock()
+	defer rn.mtx.Unlock()
+
+	rn.cmgr.DialOutboundConnection(netAddr, rn.GetId().ToUint64())
+	rn.setStatusAttempted()
+	return nil
+}
+
+// DialPersistentOutboundConnection dials a persistent outbound connection to the provided netAddr.
+func (rn *RemoteNode) DialPersistentOutboundConnection(netAddr *wire.NetAddress) error {
+	if !rn.IsNotConnected() {
+		return fmt.Errorf("RemoteNode.DialPersistentOutboundConnection: RemoteNode is not in the NotConnected state")
+	}
+
+	rn.mtx.Lock()
+	defer rn.mtx.Unlock()
+
+	rn.cmgr.DialPersistentOutboundConnection(netAddr, rn.GetId().ToUint64())
+	rn.setStatusAttempted()
+	return nil
+}
+
+// ConnectInboundPeer connects a peer once a successful inbound connection has been established.
+func (rn *RemoteNode) ConnectInboundPeer(conn net.Conn, na *wire.NetAddress) error {
```

Comment:
The comment for this function is helpful. Just a nit on the naming. Since the connection has already been established, it's actually being attached to the RemoteNode, rather than a new peer connection being formed
```suggestion
// AttachInboundConnection creates an inbound peer once a successful inbound connection has been established.
func (rn *RemoteNode) AttachInboundConnection(conn net.Conn, na *wire.NetAddress) error {
```